easybuild/easyblocks/l/llvm.py

```
@@ -381,6 +385,9 @@ def __init__(self, *args, **kwargs):
         self.log.info("Final projects to build: %s", ', '.join(self.final_projects))
         self.log.info("Final runtimes to build: %s", ', '.join(self.final_runtimes))
 
+        # CMake options passed to each build stage.
+        # Will be cleared between stages. If arguments are needed in multiple stages,
+        # consider adding them to general_opts instead.
         self._cmakeopts = {}
```

Too late for this PR now, but I thought we could set this to `None` at this point such that adding dict entries to this too early would fail loudly.

What do you think?

---

Probably a good idea. We can either add this in a separate PR, or just add it to one of the remaining LLVM ones.

---

I think https://github.com/easybuilders/easybuild-easyblocks/pull/3799 would be a good candidate as IMO that is ready.

@Crivella ?

---

I think it must only be used after the `_configure_general_build` call in `configure_step`: https://github.com/Thyre/easybuild-easyblocks/blob/9771da4f8969ba373c19b99a6988b3b65113d225/easybuild/easyblocks/l/llvm.py#L833

There cmakeopts should be initialized to a copy of `generalopts`. Or is there any earlier use intended? Maybe derived easyblocks adding options to that before calling the super-method?

---

Umh not necessarily as `_configure_general_build` does not reset them but runs an update
````
    def _configure_general_build(self):
        """General configuration step for LLVM."""
        self._cmakeopts.update(self.general_opts)
        self._add_cmake_runtime_args()
````

I agree that in general `_cmakeopts` should not be touched manually as much as possible, i would say that having it as an ~hidden attribute with the `_`  should be enough.

If we wanted to improve clarity we should add another `stage_opts` dictionary that is reset at the end of every stage and added at the start of one, but not sure it is warranted the effort and testing

---

> Right now the `_configure_...` methods are used as utilities to add to `_cmake_opts`

Hm, I would expect to have `general` always before specific `configure...` calls which is always the case except for that one instance. So I guess it would make sense to move it further below to https://github.com/Thyre/easybuild-easyblocks/blob/9771da4f8969ba373c19b99a6988b3b65113d225/easybuild/easyblocks/l/llvm.py#L833 to be in line with other usages

> The idea of the EB right now is to use `configure_step` as if it was `configure_step1` vs the `configure_step2/3` counterparts.

But that is violated by that switched order, isn't it? https://github.com/Thyre/easybuild-easyblocks/blob/9771da4f8969ba373c19b99a6988b3b65113d225/easybuild/easyblocks/l/llvm.py#L881-L883

> but i think it should be a valid variable throughout all of the configure step

Then we could unexpectedly overwrite an intentional options by the `_configure_general_build` 

> In general i would say that having `general_opts` as options used for every stage while `_cmakeopts` is valid only for the current stage is fine, i think there is a greater chance of reintroducing errors if we refactor again, vs reviewing changes in the future.

IMO there are still more surprises compared to "`_cmakeopts` must only be used after a call to `_configure_general_build` which sets it to (only) the defaults."   
I'd like to reduce them as it was missed in a change and the review before. So to me those additional safeguards do make sense

I'd do that change in a PR that needs testing against a couple ECs anyway so the chance of catching bugs by this change (even though I don't expect any) is high enough.

---

> If you set self._cmakeopts = None in __init__ it will still be an attribute, won't it?

That really only change the type of exception you would get from an `AttributeError` if it is not set to a `TypeError` if it is set to `None` and you try to assign a key to it

---

Then I dont understand your remark:

> Only if `_cmakeopts` was not an attribute there which i still think it should.

Getting an error is intended to detect using it too early. And IMO the explicit `None` shows that it does not exist yet and provides means to add a comment what it is and where it will be set.

---

What i meant is that i think `_cmakeopts` should be already usable in the configure step.

I think unless we decide do alter the logic quite a bit this will not really save us from PRs using the wrong attributes/methods if we do not catch it in review...
What i think is more worth the effort is increase the sanity checks to make sure all the feature we want on are on.

---

But should it be usable before `_configure_general_build` in the configure step?

I'd just make it as hard as possible to use the wrong one if it doesn't impact what we really need and my suggestion is a step towards it while not impacting current usage as far as I can tell.

Better sanity checks would be good either way, yes.

Then maybe the minimal common ground would be renaming the generically named `cmakeopts` to something more specific like "_stage_cmake_opts"? That would make it easier to review without any real change

---

> But should it be usable before _configure_general_build in the configure step?

I would say so, otherwise we would need a section for updating `general_opts` based on present packages (what we have right now before `_configure_general_build` in the `configure_step` and one to update the single stage after.
I feel that having 2 section there could also lead to confusion and people misplacing stuff (eg modifying `general_opts` after the call to `_configure_general_build`

> Then maybe the minimal common ground would be renaming the generically named cmakeopts to something more specific like "_stage_cmake_opts"? That would make it easier to review without any real change
Not against renaming variables for clarity, might even make it `_current_stage_cmake_opts` for clarity

I think the only way to really make sure stuff is not misused is to introduce 4 methods:
- add_stage1_cmake_opts
- add_stage2_cmake_opts
- add_stage3_cmake_opts
- add_all_stages_cmake_opts

And than we could ensure they are never called after a particular stage has already been ran, and that the internal methods are never touched manually.
This would require quite a lot of refactoring, and high possibility to introduce new errors which i would avoid for now